src/Microsoft.AspNetCore.Mvc.Core/Internal/ResponseCacheFilter.cs

```diff
@@ -110,6 +121,16 @@ public void OnActionExecuting(ActionExecutingContext context)
                 headers[HeaderNames.Vary] = VaryByHeader;
             }
 
+            if (VaryByQueryKeys != null)
+            {
+                var responseCacheFeature = context.HttpContext.Features.Get<IResponseCacheFeature>();
+                if (responseCacheFeature == null)
+                {
+                    throw new InvalidOperationException($"The response cache middleware must be added when using the {nameof(VaryByQueryKeys)} feature.");
+                }
+                responseCacheFeature.VaryByQueryKeys = VaryByQueryKeys;
```

Comment:
I thought this type was primarily for browser caching and not server side caching. Are we repurposing it?